README.md
[![Swift 5.5](https://img.shields.io/badge/Swift-5.5-orange.svg?style=flat)](https://swift.org)
[![Platforms iOS](https://img.shields.io/badge/Platforms-iOS-lightgray.svg?style=flat)](http://www.apple.com)
[![CocoaPods](https://img.shields.io/cocoapods/v/AEConsole.svg?style=flat)](https://cocoapods.org/pods/AEConsole)
[![Carthage](https://img.shields.io/badge/Carthage-compatible-brightgreen.svg?style=flat)](https://github.com/Carthage/Carthage)
[![Swift Package Manager](https://img.shields.io/badge/SPM-compatible-brightgreen.svg)](https://github.com/apple/swift-package-manager)
[![License MIT](https://img.shields.io/badge/License-MIT-lightgrey.svg?style=flat)](https://github.com/tadija/AELog/blob/master/LICENSE)

# AEConsole

**Customizable Console UI overlay with debug log on top of your iOS App**

**AEConsole** is built on top of [AELog](https://github.com/tadija/AELog), so you should probably see that first.
> I wanted to see what's happening 'under the hood' while testing some app AFK (ex. outside).
> Then I made it possible. Hope you'll like it too, happy coding!

![AEConsole](http://tadija.net/projects/AEConsole/AEConsole.png)

## Index
- [Features](#features)
- [Usage](#usage)
- [Quick Help](#quick-help)
- [Installation](#installation)
- [License](#license)

## Features
- **All the things** from [AELog](https://github.com/tadija/AELog) plus:
- Console UI overlay **on top of your App**
- **Debug log on device** in real time
- **Automatic row height** for all log lines
- **Forward touches** to your App
- **Shake to toggle** Console UI
- **Filter log** to find exactly what you need
- **Export log** to file if you need it for later
- **Share log file** easily via system sharing sheet
- **Customize look** as you like it

## Usage

Calling `Console.shared.configure(in: window)` will add `Console.View` as a subview to your App's window and make it hidden by default.
Whenever you need Console UI, you just make a shake gesture and it's there! When you no longer need it, shake again and it's gone.

The rest is up to [AELog's](https://github.com/tadija/AELog) logging functionality. Whatever is logged with it, will show up in `Console.View`.

```swift
// MARK: - Console configuration

func application(_ application: UIApplication,
                 didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool
{
    /// - Note: Access Console settings
    let settings = Console.shared.settings

    /// - Note: Customize Console settings like this, these are defaults:
    settings.isShakeGestureEnabled = true
    settings.backColor = UIColor.black
    settings.textColor = UIColor.white
    settings.fontSize = 12.0
    settings.rowSpacing = 4.0
    settings.opacity = 0.7

    /// - Note: Configure Console in app window (it's recommended to skip this for public release)
    Console.shared.configure(in: window)

    /// - Note: Log something with AELog
    aelog()

    return true
}
```

```swift
// MARK: - Additional Console API

/// - Note: Check if Console is hidden
Console.shared.isHidden

/// - Note: Toggle Console visibility
Console.shared.toggle()

/// - Note: Add any log line manually
Console.shared.addLogLine(line: "Hello!")

/// - Note: Export log file manually
Console.shared.exportLogFile { (fileURL) in
    do {
        let url = try fileURL()
        /// - Note: do something with a log file at given file URL...
    } catch {
        print(error)
    }
}
```

## Quick Help

This should explain all the features of Console UI:

![AEConsole](http://tadija.net/projects/AEConsole/AEConsole-QuickHelp.png)

Feature | Description
------------ | -------------
Export Log | will make `AELog_{timestamp}.txt` file in Application Documents directory and present sharing sheet.
Filter Log | filter is not case sensitive.
Toggle Toolbar | works for both filter and menu toolbars simultaneously.
Toggle Forward Touches | when active you can interact with your App, otherwise you can interact with the log.
Toggle Auto Follow | when active it will automatically scroll to the new log lines, otherwise it will stay put.
Clear Log | you can't undo this.
Pan Gesture over Menu Toolbar | left is more transparent, right is more opaque.

## Installation

- [Swift Package Manager](https://swift.org/package-manager/):
    
    ```swift
    .package(url: "https://github.com/tadija/AEConsole.git", from: "0.7.2")
    ```
  
- [Carthage](https://github.com/Carthage/Carthage):

	```ogdl
	github "tadija/AEConsole"
	```
	
- [CocoaPods](http://cocoapods.org/):

	```ruby
	pod 'AEConsole'
	```

## License
AEConsole is released under the MIT license. See [LICENSE](LICENSE) for details.


CHANGELOG.md
# Changelog

## Version 0.7.2
- Upgrade project to recommended settings with Xcode 13.3
- Updated AELog dependency to 0.6.3

## Version 0.7.1
- Minor changes

## Version 0.7.0
- Migrate to Swift 5 with Xcode 11 (11A420a)
- Update "swift_versions" in podspec file

## Version 0.6.1
- Bumped Swift language version to 4.2 with Xcode 10 (10A255)
- Reorganizing project directories and files

## Version 0.6.0
- Added sharing of exported log file
- Automatic row height for all log lines
- Replaced `rowHeight` with `rowSpacing` in `Settings`
- Other improvements and refactoring

## Version 0.5.5
- Updated AELog dependency to 0.5.5

## Version 0.5.0
- Fixed issues #2, #3, #4, #5
- Improvements and refactoring
- Updated AELog dependency to 0.5.1

## Version 0.4.1
- Added support for iPhone X
- Improvements and bug fixes

## Version 0.4.0
- Migrated to Swift 4 and recommended settings with Xcode 9.2
- Replaced `Config` (from .plist file) with `Settings` class (configure via code)
- Renamed class `AEConsole` to `Console`

## Version 0.3.1
- Minor fixes

## Version 0.3.0
- Fixed all build warnings and errors in Xcode 8.1 Beta (8T47)
- Code updated for Swift 3.0 ("swiftyfying")
- Sources reorganized from single to multiple files

## Version 0.2.2
- Fixed bug when logging from background thread
- Major performance improvement

## Version 0.2.1
- Ready for release

## Version 0.1.0
- Initial version


Sources/Config.swift
/**
 *  https://github.com/tadija/AEConsole
 *  Copyright (c) Marko Tadić 2016-2018
 *  Licensed under the MIT license. See LICENSE file.
 */

import UIKit

/**
    Helper for accessing settings from the external file.

    Create `AEConsole.plist` dictionary file and add it to your target.
    Alternative is to add `AEConsole` dictionary inside existing `Info.plist` file.

    There is `Key` struct which contains possible keys for all settings.
*/
public class Config {
    
    // MARK: - Constants
    
    /// Setting keys which can be used in `AEConsole` dictionary.
    public struct Key {
        /// Boolean - Console UI enabled flag (defaults to `NO`)
        public static let Enabled = "Enabled"
        
        /// Boolean - Console UI visible on app start flag (defaults to `NO`)
        public static let AutoStart = "AutoStart"
        
        /// Boolean - Shake gesture enabled flag (defaults to `YES`)
        public static let ShakeGesture = "ShakeGesture"
        
        /// String - Hex string for Console background color (defaults to 000000)
        public static let BackColor = "BackColor"
        
        /// String - Hex string for Console text color (defaults to FFFFFF)
        public static let TextColor = "TextColor"
        
        /// Number - Console UI font size (defaults to 12)
        public static let FontSize = "FontSize"
        
        /// Number - Console UI row height (defaults to 14)
        public static let RowHeight = "RowHeight"
        
        /// Number - Console UI opacity (defaults to 0.7)
        public static let Opacity = "Opacity"
    }
    
    private struct Default {
        static let Enabled = false
        static let AutoStart = false
        static let ShakeGesture = true
        static let BackColor = UIColor.black
        static let TextColor = UIColor.white
        static let FontSize: CGFloat = 12.0
        static let RowHeight: CGFloat = 14.0
        static let Opacity: CGFloat = 0.7
    }
    
    // MARK: - Properties
    
    static let shared = Config()
    
    lazy var textColorWithOpacity: UIColor = { [unowned self] in
        self.textColor.withAlphaComponent(Default.Opacity)
    }()
    
    /// Contents of AEConsole settings (AEConsole.plist file or AEConsole dictionary from Info.plist)
    private lazy var data: [String : AnyObject]? = {
        guard let
            path = Bundle.main.path(forResource: "AEConsole", ofType: "plist"),
            let data = NSDictionary(contentsOfFile: path) as? [String : AnyObject]
        else { return self.alternateData }
        return data
    }()
    
    private lazy var alternateData: [String : AnyObject]? = {
        guard let data = Bundle.main.infoDictionary?["AEConsole"] as? [String : AnyObject]
        else { return nil }
        return data
    }()
    
    // MARK: - Settings
    
    lazy var consoleFont: UIFont = {
        return UIFont.monospacedDigitSystemFont(ofSize: self.fontSize, weight: UIFont.Weight.regular)
    }()
    
    lazy var isEnabled: Bool = { [unowned self] in
        guard let enabled = self.getBool(with: Key.Enabled)
        else { return Default.Enabled }
        return enabled
    }()
    
    lazy var isAutoStartEnabled: Bool = { [unowned self] in
        guard let autoStart = self.getBool(with: Key.AutoStart)
        else { return Default.AutoStart }
        return autoStart
    }()
    
    lazy var isShakeGestureEnabled: Bool = { [unowned self] in
        guard let shake = self.getBool(with: Key.ShakeGesture)
        else { return Default.ShakeGesture }
        return shake
    }()
    
    lazy var backColor: UIColor = { [unowned self] in
        guard let color = self.getColor(with: Key.BackColor)
        else { return Default.BackColor }
        return color
    }()
    
    lazy var textColor: UIColor = { [unowned self] in
        guard let color = self.getColor(with: Key.TextColor)
        else { return Default.TextColor }
        return color
    }()
    
    lazy var fontSize: CGFloat = { [unowned self] in
        guard let fontSize = self.getNumber(with: Key.FontSize)
        else { return Default.FontSize }
        return fontSize
    }()
    
    lazy var rowHeight: CGFloat = { [unowned self] in
        guard let rowHeight = self.getNumber(with: Key.RowHeight)
        else { return Default.RowHeight }
        return rowHeight
    }()
    
    lazy var opacity: CGFloat = { [unowned self] in
        guard let opacity = self.getNumber(with: Key.Opacity)
        else { return Default.Opacity }
        return opacity
    }()
    
    // MARK: - Helpers
    
    private func getBool(with key: String) -> Bool? {
        guard let
            data = self.data,
            let bool = data[key] as? Bool
        else { return nil }
        return bool
    }
    
    private func getNumber(with key: String) -> CGFloat? {
        guard let
            data = self.data,
            let number = data[key] as? CGFloat
        else { return nil }
        return number
    }
    
    private func getColor(with key: String) -> UIColor? {
        guard let
            data = self.data,
            let hex = data[key] as? String
        else { return nil }
        let color = getColor(from: hex)
        return color
    }
    
    private func getColor(from hex: String) -> UIColor? {
        let scanner = Scanner(string: hex)
        var hexValue: UInt32 = 0
        if scanner.scanHexInt32(&hexValue) {
            let red   = CGFloat((hexValue & 0xFF0000) >> 16) / 255.0
            let green = CGFloat((hexValue & 0x00FF00) >> 8) / 255.0
            let blue  = CGFloat((hexValue & 0x0000FF)) / 255.0
            let color = UIColor(red: red, green: green, blue: blue, alpha: 1.0)
            return color
        } else { return nil }
    }
    
}
